commcare_connect/opportunity/models.py

```diff
@@ -75,46 +78,95 @@ class Opportunity(BaseModel):
+                .distinct()
+            )
+        )
+
+    @property
+    def minimum_budget_per_visit(self):
+        return min(self.top_level_paymentunits.values_list("amount", flat=True))
+
     @property
     def remaining_budget(self) -> int:
         return self.total_budget - self.claimed_budget
 
     @property
     def claimed_budget(self):
-        return self.claimed_visits * self.budget_per_visit
+        opp_access = OpportunityAccess.objects.filter(opportunity=self)
+        opportunity_claim = OpportunityClaim.objects.filter(opportunity_access__in=opp_access)
+        claim_limits = OpportunityClaimLimit.objects.filter(opportunity_claim__in=opportunity_claim)
+
+        payment_unit_counts = claim_limits.values("payment_unit").annotate(
+            visits_count=Sum("max_visits"), amount=F("payment_unit__amount")
+        )
+        claimed = 0
+        for count in payment_unit_counts:
+            visits_count = count["visits_count"]
+            amount = count["amount"]
+            claimed += visits_count * amount
+
+        return claimed
 
     @property
     def utilised_budget(self):
-        return self.approved_visits * self.budget_per_visit
+        completed_works = CompletedWork.objects.filter(opportunity_access__opportunity=self)
+        payment_unit_counts = completed_works.values("payment_unit").annotate(
+            completed_count=Count("id"), amount=F("payment_unit__amount")
+        )
+        utilised = 0
+        for payment_unit_count in payment_unit_counts:
+            completed_count = payment_unit_count["completed_count"]
+            amount = payment_unit_count["amount"]
+            utilised += completed_count * amount
+        return utilised
 
     @property
     def claimed_visits(self):
         opp_access = OpportunityAccess.objects.filter(opportunity=self)
-        used_budget = OpportunityClaim.objects.filter(opportunity_access__in=opp_access).aggregate(
-            Sum("max_payments")
-        )["max_payments__sum"]
+        opportunity_claim = OpportunityClaim.objects.filter(opportunity_access__in=opp_access)
+        used_budget = OpportunityClaimLimit.objects.filter(opportunity_claim__in=opportunity_claim).aggregate(
+            Sum("max_visits")
+        )["max_visits__sum"]
         if used_budget is None:
             used_budget = 0
         return used_budget
 
     @property
     def approved_visits(self):
-        approved_user_visits = UserVisit.objects.filter(
-            opportunity=self, status=VisitValidationStatus.approved
-        ).count()
-        return approved_user_visits
+        return CompletedWork.objects.filter(opportunity_access__opportunity=self).count()
+
+    @property
+    def number_of_users(self):
+        return self.total_budget / self.budget_per_user
 
     @property
     def allotted_visits(self):
-        return math.floor(self.total_budget / self.budget_per_visit)
+        payment_units = self.top_level_paymentunits.all()
```

Comment:
why do all of these use `top_level_paymentunits`

Reply:
This came after a discussion I had with Rama. The gist of it was that the payment-units which are prerequisites (the ones other than top-level paymentunits) would not be considered for budget calculations and any payment-amount you set for that is ignored. 

```
[10:28 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997106085829)
Sravan Reddy
 Per https://dimagi.atlassian.net/browse/CCCT-276, we are not having per user limits any more, is it?

[CCCT-276 D2ORG - Opportunity Setup Changes](https://dimagi.atlassian.net/browse/CCCT-276?atlOrigin=eyJpIjoiMDhmZmFkZjQzNDYwNDBiNmI0MTRhYjkzMzUyZTk1NmQiLCJwIjoiamlyYS1zbGFjay1pbnQifQ)
Status: In Progress
Type: Story
Assignee: Sravan Reddy
Priority: P6
More actions...
Added by [Jira Cloud](https://dimagi.slack.com/services/B027GSXHAKD)
[10:28 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997119623709)
From 2.e and 2.f
Allow me to set a limit on the number of deliveries that are possible in total for that unit of work
Allow me to set a daily limit on the number of deliveries that are possible on any day before the Opportunity end date
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997142149099)
Rama Siva
 Let me check
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997143298559)
Sravan Reddy
 Or is that per user?
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997149409929)
Okay
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997152394819)
Rama Siva
 This has always been per user
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997176018529)
So no real change here other than making it per 'payment unit'
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997188441789)
Sravan Reddy
 I see, makes sense
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997193260479)
Rama Siva
 cool
[10:29 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997199850309)
Sravan Reddy
 Thanks for clarification.
[10:30 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997213800019)
Rama Siva
 :+1:
[10:33 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997437712669)
Sravan Reddy
 How do we change this section since it’s per Payment Unit now?
Screenshot 2024-03-21 at 10.33.37 AM.png
 
Screenshot 2024-03-21 at 10.33.37 AM.png
[10:34 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997461426479)
Meaning total visits now would be - X number of PU1 and Y number of PU2 (PU= payment unit)
[10:35 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997517347399)
Sravan Reddy
 Shall we just say X+Y?

1 reply
15 days agoView thread
[10:35 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997533763029)
Rama Siva
 great point. a 'visit' in this sense would map to a payment unit. so you could sum up the daily limit for all payment units
[10:35 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997550535169)
Sravan Reddy
 Cool
[10:36 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997576087669)
Rama Siva
 We do need to clean up our language in subsequent builds
[10:37 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997646311979)
Sravan Reddy
 Yeah
[10:37 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997678639769)
Rama Siva
 But in the case of sub-units of work this will not work.
[10:38 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997701665529)
So it will need to be the top level limit
[10:38 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710997723119449)
because we are really counting the number of beneficiaries in a day
[10:43 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998003210099)
Sravan Reddy
 Right, yeah
[10:43 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998009946669)
We will just do the top level limits
[10:44 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998064542939)
Rama Siva
 So the user can do 2 verified service deliveries and that is the limit shown here even if the sub-unit of work (verified screening) limit is set to 3
[10:45 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998116097949)
In reality, it is going to be the same and we should debate later if allowing a separate limit for a sub-unit of work is needed
[10:45 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998121338339)
Sravan Reddy
 So, we don’t take  the sub-units of work into account while calculating limits?
[10:45 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998129467199)
Yeah
[10:45 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998139458489)
Rama Siva
 Yeah, that is what I meant
[10:45 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998146013939)
Sravan Reddy
 Right now, the UI does allow specifying limits for sub unit of work as well


2 replies
Last reply 15 days agoView thread
[10:46 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998160109629)
Rama Siva
 they should just know how many verified service deliveries they should be doing
[10:46 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998209062339)
If the UI does not allow setting a separate limit then we should not do any addition
[10:47 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998235962019)
But if it is not a sub-unit, we do allow a separate limit?
[10:47 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998238945499)
on the UI?
[10:47 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998253583709?thread_ts=1710998146.013939&cid=D057T7G55DJ)
Sravan Reddy
 
replied to a thread:
Right now, the UI does allow specifying limits for sub unit of work as well
It does, but we can ignore them for calculating limits and budget, and let project-user know  that limits won’t be applicable if a payment-unit is made subunit of another payment unit
[10:47 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998268346029)
All, the paymentunits have a limit
[10:48 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998280196519)
Some payment-units can have sub-payment-units
[10:48 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998301825139)
Rama Siva
 great. i think we are aligned
[10:48 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998317183439)
Sravan Reddy
 Cool
[10:48 AM](https://dimagi.slack.com/archives/D057T7G55DJ/p1710998334325489)
Rama Siva
 great catch
```

Reply:
We chatted about this in the weekly meeting and I think there was a mis alignment somewhere previously, but we decided as a group that all payment units should contribute to the budget and visit counts, since for many projects (even if not all of them), the payment unit dependencies are true independent visits happening at different times that accrue different payments, and even for the chc use case here, the accrual happens in the "sub unit" not the top level one.